physicalplan/plan.go

```diff
@@ -1,48 +1,69 @@
 // Copyright (c) The Thanos Community Authors.
 // Licensed under the Apache License 2.0.
 
+// Copyright 2013 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package physicalplan
 
 import (
 	"runtime"
 	"time"
 
+	"github.com/efficientgo/core/errors"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/storage"
+
 	"github.com/thanos-community/promql-engine/physicalplan/aggregate"
 	"github.com/thanos-community/promql-engine/physicalplan/binary"
 	"github.com/thanos-community/promql-engine/physicalplan/exchange"
 	"github.com/thanos-community/promql-engine/physicalplan/model"
 	"github.com/thanos-community/promql-engine/physicalplan/parse"
 	"github.com/thanos-community/promql-engine/physicalplan/scan"
+	"github.com/thanos-community/promql-engine/physicalplan/step_invariant"
 	"github.com/thanos-community/promql-engine/physicalplan/unary"
-
-	"github.com/efficientgo/core/errors"
-	"github.com/prometheus/prometheus/promql/parser"
-	"github.com/prometheus/prometheus/storage"
 )
 
 const stepsBatch = 10
 
 // New creates new physical query execution plan for a given query expression.
 func New(expr parser.Expr, storage storage.Queryable, mint, maxt time.Time, step, lookbackDelta time.Duration) (model.VectorOperator, error) {
-	return newCancellableOperator(expr, storage, mint, maxt, step, lookbackDelta)
+	// Pre-process the expression to check whether the
+	// expression is step invariant.
+	expr = promql.PreprocessExpr(expr, mint, maxt)
```

Comment:
I wonder if we should add this to the logical plan after https://github.com/thanos-community/promql-engine/pull/79?

Reply:
Yeah that's a great idea.
If we have logical plan then dealing with this should be much more easier.